src/zenml/integrations/modal/orchestrators/modal_sandbox_executor.py

```diff
@@ -0,0 +1,683 @@
+            )
+            return self.shared_image_cache[cache_key]
+
+        # Fallback to existing image building logic
+        logger.debug(
+            f"Building new Modal image for {step_name or 'pipeline'}: {image_name}"
+        )
+        if self.deployment.build is None:
+            raise ValueError("Deployment build is None, cannot build image")
+
+        return get_or_build_modal_image(
+            image_name=image_name,
+            stack=self.stack,
+            pipeline_name=self.deployment.pipeline_configuration.name,
+            build_id=str(self.deployment.build.id),
+            app=self.app,
+        )
+
+    def _get_image_cache_key(
+        self, image_name: str, step_name: Optional[str] = None
+    ) -> str:
+        """Generate a cache key for Modal images.
+
+        Args:
+            image_name: The base image name.
+            step_name: Name of the step (None for pipeline-level).
+
+        Returns:
+            Cache key for the image.
+
+        Raises:
+            ValueError: If the deployment does not have a build ID which is
+                required to scope the cache key.
+        """
+        # Use build ID and step name to create unique cache key
+        # Include a hash of the image name for uniqueness
+        if self.deployment.build is None:
+            raise ValueError(
+                "Deployment build is None, cannot generate cache key"
+            )
+
+        build_id = str(self.deployment.build.id)
+        image_hash = str(hash(image_name))[-8:]  # Last 8 chars of hash
+        if step_name:
+            return f"{build_id}_{step_name}_{image_hash}"
+        else:
+            return f"{build_id}_pipeline_{image_hash}"
+
+    async def execute_pipeline(
+        self,
+        run_id: Optional[str] = None,
+        synchronous: bool = True,
+    ) -> None:
+        """Execute the entire pipeline in a single sandbox.
+
+        Args:
+            run_id: The pipeline run ID.
+            synchronous: Whether to wait for completion.
+        """
+        logger.debug("Executing entire pipeline in single sandbox")
+
+        # Build entrypoint command
+        command = (
+            ModalOrchestratorEntrypointConfiguration.get_entrypoint_command()
+        )
+        from uuid import UUID
+
+        # Convert run_id to UUID if it's a string
+        run_id_uuid = None
+        if run_id is not None:
+            run_id_uuid = UUID(run_id) if isinstance(run_id, str) else run_id
+
+        args = (
+            ModalOrchestratorEntrypointConfiguration.get_entrypoint_arguments(
+                deployment_id=self.deployment.id,
+                run_id=run_id_uuid,
+            )
+        )
+        entrypoint_command = self._build_entrypoint_command(command, args)
+
+        # Execute pipeline sandbox
+        await self._execute_sandbox(
+            entrypoint_command=entrypoint_command,
+            mode="PIPELINE",
```

Comment:
Actually the mode here can be either pipeline or step, because this launches a sandbox that either
- runs the entire pipeline
- only orchestrator the steps
So we should rather pass the actual mode here, or nothing at all. This is anyway only used for tags.